lib/queue.js

```diff
@@ -726,65 +725,27 @@ Queue.prototype.processJob = function(job){
 
     stopTimer();
 
-    // TODO: Should moveToFailed ensure the lock atomically in one of its Lua scripts?
     // See https://github.com/OptimalBits/bull/pull/415#issuecomment-269744735
-    job.takeLock(true /* renew */, false /* ensureActive */).then( function(/*lock*/) {
-      return job.moveToFailed(err).finally(function(){
-        _this.processing--;
-      }).then(function(){
-        return _this.distEmit('failed', job, error);
-      });
-    }, function(err){
-      _this.emit('error', err, 'failed to re-obtain lock before moving to failed, bailing');
+    //  job.takeLock(true /* renew */, false /* ensureActive */).then( function(/*lock*/) {
+    return job.moveToFailed(err).then(function(){
```

Comment:
I think there's a hazard here: the lock can expire immediately after `stopTimer` is called and before the job is moved to failed. If `moveUnlockedJobsToWait` happens to run then, it'll reprocess the job. That's why we previously called `job.takeLock` right before calling job.moveToFailed.

To fix this, I think `stopTimer` should be moved to be called in a `.finally()` block after `moveToFailed`. That way `moveToFailed` can assume it has a lock.